### PR TITLE
[rllib] [WIP] RND

### PR DESCRIPTION
**Pull request by visatish**
WIP RND.

### CODE REVIEW

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 1]
```diff
@@ -212,6 +212,8 @@ def to_batches(tensor):
             self,
             observation_space,
             action_space,
+            config,
+            logit_dim,
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: Consider using a fixed-length vector here (e.g., rnd_num_units).

File: python/ray/rllib/agents/impala/impala.py [hunk 2]
```diff
@@ -71,6 +71,8 @@
     "max_sample_requests_in_flight_per_worker": 2,
     # max number of workers to broadcast one set of weights to
     "broadcast_interval": 1,
+    # whether or not to enable exploration through Random Network Distillation (RND)
+    "rnd": 0,
```
Comment (ericl) [inline python/ray/rllib/agents/impala/impala.py]: Move to agent.py

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 3]
```diff
@@ -212,6 +212,8 @@ def to_batches(tensor):
             self,
             observation_space,
             action_space,
+            config,
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: Pass it in for the other algorithms as well

**Comment by ericl** [comment]
You should also check in your unit test benchmarks (put under rllib/test).

File: python/ray/rllib/utils/rnd.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        return intr_rew
+
+    def compute_intr_rew(self, obs):
+        return np.squeeze(self._sess.run(self._intr_rew, feed_dict={self._obs_ph: obs, self._is_training_ph: False}))
+
+    def _build_loss(self):
+        logger.info('Building RND loss...')
+        loss = tf.reduce_mean(tf.square(self._preds - self._targets), axis=-1)
+        keep_mask = tf.random_uniform(shape=tf.shape(loss), minval=0.0, maxval=1.0, dtype=tf.float32)
+        keep_mask = tf.cast(keep_mask < self._rnd_predictor_update_proportion, tf.float32)
```
Comment (ericl) [inline python/ray/rllib/utils/rnd.py]: It's kind of weird to randomly drop out the loss, is this right?

**Comment by ericl** [comment]
[auto-closing stale pr; feel free to re-open]
